### ISSUE TITLE
Variadic map function

### ISSUE THREAD

**Issue by njordhov** [clarity-lang/reference#27]
The [map](https://docs.blockstack.org/core/smart/clarityref#map) function in Clarity currently is limited to functions that takes a single argument, mapping them over a single sequence.

The `map` function should be variadic and allow functions that takes more than one argument, mapping them over a number of sequences matching their arity. 

For example:

```clojure
(map + (list 1 2 3) (list 4 5 6))     ;; evaluates to (5 7 9) 
```

Variadic `map` functions are found in similar other languages including [Racket](https://docs.racket-lang.org/reference/pairs.html#%28def._%28%28lib._racket%2Fprivate%2Fmap..rkt%29._map%29%29), [Scheme](https://www.gnu.org/software/mit-scheme/documentation/mit-scheme-ref/Mapping-of-Lists.html), [Common Lisp](http://clhs.lisp.se/Body/f_map.htm) and [Clojure](https://clojuredocs.org/clojure.core/map).

Originally posted at: https://github.com/blockstack/stacks-blockchain/issues/1482

**Comment by njordhov** [comment]
In functional programming, it is commonplace to use `map` with _partial function applications_ and _anonymous functions_, typically closing over local bindings.  **Variadic map could compensate for Clarity lacking closures** by providing an alternative means to pass arguments from the local context to the mapping function.

With multiple sequences as arguments, it's an open question how to handle lists of different length. [Racket](https://docs.racket-lang.org/reference/pairs.html#%28def._%28%28lib._racket%2Fprivate%2Fmap..rkt%29._map%29%29) takes a strict approach, requiring all input lists to have the same length. In contrast, [Common Lisp](http://clhs.lisp.se/Body/f_map.htm) and [Clojure](https://clojuredocs.org/clojure.core/map) only iterate to the end of the shortest input sequence.  

**Clarity can benefit from variadic `map` implicitly extending shorter input sequences to the length of the longest** by repeating their last item. This can be used to pass on arguments to the applied function, without requiring an explicit expression such as [`repeat`](https://clojuredocs.org/clojure.core/repeat) to fill the shorter sequences. 

Use case from the Blockstack codebase:
https://github.com/blockstack/stacks-blockchain/blob/27381d64b1e6c8f308537a2c397958a97fdea337/src/chainstate/stacks/boot/pox.clar#L338-L352

Assuming that `map` fills list arguments to have the same length, this could be an alternative implementation:

```clojure
(fold + 
  (map add-pox-addr-to-ith-reward-cycle 
    (list {pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx})
    (list u0 u1 u2 u3 u4 u5 u6 u7 u8 u9})))
```

It is of course possible to rather use `fold` to accumulate the integers, or even combining into a single reducer. A benefit of using `map` for such mapping is that the applied function will be simpler and more straight-forward, returning a single integer result for each iteration with the accumulated result immediately available, while `fold` has to pass on arguments in the accumulator and requires extracting the accumulated result on completion.

**Comment by njordhov** [comment]
Variadic map can substantially simplify the implementation of various sequence processing. For example, just testing whether an item is in a list takes a good chunk of Clarity code. With variadic `map` it can be condensed into a one-liner:

```clojure
(fold or (map is-eq (list item) items) false)    ;; true if the item is in the items list
```
Compare to the implementation of `contains` and `contains-check` in the [rockets-base.clar](https://docs.blockstack.org/smart-contracts/principals#example-authorization-checks) example:

```clojure
;; implementing a contains function via fold
(define-private (contains-check
                   (y principal)
                   (to-check {p: principal, result: bool}))
   (if (get result to-check)
        to-check
        {p: (get p to-check),
         result: (is-eq (get p to-check) y)}))

(define-private (contains (item principal) (items (list 10 principal)))
   (get result (fold contains-check items
                    {p: item, result: false)))
```

**Comment by kantai** [comment]
Variadic map was implemented in Clarity in Stacks 2.0
